### ISSUE TITLE
Watcher getting unresponsive (internal crash) after providing invalid path to image

### ISSUE THREAD

**Issue by Andrew15-5** [typst/typst#2633]
### Description

I have "save on each keystroke" and every single time (99%) when I want to write something like:

```typ
#image("") // 1
#image("./") // 2
#image("./dir") // 3
#image("./dir/") // 4
#image("./dir/file.png") // 5
```

The rendering works on step 1 and 2, but _sometime_ (depending on how you combine manual typing + autocompletion the rest of the steps are handled fine (errors are shown and program keeps working), but sometimes it either shows error "file not found" or "? is a directory" (while internally crashing) or just shows last successful rendering and no error (while also internally crashing).

### Operating system

Linux

### Typst version

- [X] I am using the latest version of Typst

**Comment by laurmaedje** [comment]
Possibly the same issue as https://github.com/typst/typst/issues/2601

**Comment by Andrew15-5** [comment]
Yes, this may be the same problem. I need to conduct more tests.

**Comment by laurmaedje** [comment]
I have created https://github.com/typst/typst/pull/2706 as an attempt at fixing #2601. It would be great if you could test whether it resolves this issue.

**Comment by Andrew15-5** [comment]
Nope, the watcher freezes in v0.8, v0.9 and in 44d0e9c5beba6d285c6d79027d4e28a8331b260d.

**Comment by Andrew15-5** [comment]
Sometime after I insert `images` from suggested `images/` item it stops responding, and other time it stops responding after I manually write `ima`.

**Comment by laurmaedje** [comment]
Does "internal crash" mean you need to restart typst watch for it to do anything at all after that?

**Comment by Andrew15-5** [comment]
Yes.

**Comment by laurmaedje** [comment]
Which editor are you using and does it also happen if you disable auto-save but save manually quickly after the edits?

**Comment by Andrew15-5** [comment]
> Which editor are you using

Neovim

> does it also happen if you disable auto-save but save manually quickly after the edits?

Yes. I don't need to and can't physically save "quickly". That's the auto-save plugin's job.

**Comment by holsta** [comment]
I can reproduce this behaviour with Neovim on a new build on OpenBSD:

````
typst -vvvv watch foo.typ
[...]

watching foo.typ
writing to foo.pdf

[15:50:21] compiled successfully in 10.67ms

TRACE mio::poll: registering event source with poller: token=Token(0), interests=READABLE
 INFO watch_dependencies: typst::watch: Watching /home/holsta/3rdparty/typst/foo.typ
TRACE notify::kqueue: adding kqueue watch: /home/holsta/3rdparty/typst/foo.typ
````
Saving once from Neovim:
kkkk

**Comment by Andrew15-5** [comment]
@laurmaedje, I've created a Docker + Docker Compose + Just reproduction bundle in the OP. You literally only need to run `j` (`just`), open Typst logs in a new terminal and press `a` in the Neovim and the Typst will become unresponsive.

**Comment by laurmaedje** [comment]
> I've created a Docker + Docker Compose + Just reproduction bundle in the OP.

Thanks, I will take a look at it tomorrow.

**Comment by Andrew15-5** [comment]
@laurmaedje, were you able to test _de thing_ (reproduction)?

**Comment by laurmaedje** [comment]
Unfortunately, I got this error:
```
 > [neovim 6/8] RUN nvim:
0.081 /usr/local/bin/nvim: 2: nvim.appimage: Exec format error
```
I'm not sure how to fix that. I'm running on macOS, but would have expected docker to handle that transparently. Not sure whether it is related to that.

**Comment by laurmaedje** [comment]
I can reproduce it. My first guess would be that due to neovim moving and creating files rather than just overwriting them, Typst somehow ends up unwatching the main file.

**Comment by Andrew15-5** [comment]
Wait what? Moving and creating? Let me check...

`inotifywait -mr .`:

```
auto-save:
./ OPEN a.typ
./ MODIFY a.typ
./ MODIFY a.typ
./ CLOSE_WRITE,CLOSE a.typ
./ ATTRIB a.typ

manual:
./ OPEN a.typ
./ MODIFY a.typ
./ CLOSE_WRITE,CLOSE a.typ
./ ATTRIB a.typ
```

It also watches dot files. First time, auto-save also produced only one MODIFY event.

Here is the `mv file1 file2`:

```
./ MOVED_FROM file1
./ MOVED_TO file2
./ CLOSE_NOWRITE,CLOSE,ISDIR 
./ OPEN,ISDIR 
./ CLOSE_NOWRITE,CLOSE,ISDIR 
```

And here is manual save of uncreated file:

```
./ CREATE file1
./ OPEN file1
./ MODIFY file1
./ CLOSE_WRITE,CLOSE file1
```

So there is clearly no moving and no creating while saving.

> rather than just overwriting them

It does override, because the very basic stuff (like modifying and saving a file) were already polished.

**Comment by Andrew15-5** [comment]
The workaround (for me for now) is to just make a valid path name string on the commented line and then just uncomment it and it should work.

```typ
#image("")
// #image("./")
// #image("./dir/img.svg")
#image("./dir/img.svg")
```

**Comment by Andrew15-5** [comment]
Another hack is to `ln -sr a.typ link.typ` and then **edit the `link.typ`** and watch either one of the files.

**Comment by Andrew15-5** [comment]
I've spoken with some Neovim devs and the problem is with how Neovim behaves during saving (`:h backupcopy`, `:h backupdir`, `:h backupskip`). The problem is probably with the `MOVED_TO` event. But if the less efficient saving behavior is set via

```vim
setlocal backupcopy=yes
```

or 

```lua
vim.opt_local.backupcopy = "yes"
```

Then the `MOVED_TO` event doesn't happen and Typst doesn't stop watching.

The problem is that this even (or just the default saving behavior) happens on every and any keystroke, including when typing `#image("./")`, but in this case Typst still watches for changes. And only when you are providing the "real" path to an image (meaning any path, even non-existing one) the Typst stops watching, and it has to be manually restarted.

In other words, Typst still _should_ work just fine with the default Neovim saving behavior, but _sometimes_ it doesn't (when searching for an image, for instance).

<details><summary>my (hopefully temporary) fix</summary>
<p>

```lua
local typst_group = vim.api.nvim_create_augroup("typst_group", {})
vim.api.nvim_create_autocmd({ "FileType" }, {
  pattern = "typst",
  callback = function()
    vim.opt_local.backupcopy = "yes"
  end,
  group = typst_group,
})
```

</p>
</details>

P.S. You can check out the conversation starting from this message: https://matrix.to/#/!cylwlNXSwagQmZSkzs:matrix.org/$fWk2CIitaD-d006OzEekO7Ql-FISil8eC7K2qhDUTVU.

**Comment by Andrew15-5** [comment]
I haven't looked at the code, but it's indeed now fixed in https://github.com/DioxusLabs/dioxus/issues/2312#issuecomment-2439079194. But I only tried it for a little.
